### PR TITLE
Enable thread cancellation to resolve OvercommitTracker deadlocks in MetadataStorageFromDisk

### PR DESCRIPTION
**Pull request by serxa**
Should be merged after https://github.com/ClickHouse/ClickHouse/pull/45007

Deadlocks are possible in the first place due to the non-deterministic locking order. Namely when OvercommitTracker tries to cancel a query while holding a lock (e.g. MetadataStorageFromDisk::metadata_mutex), but one of the threads of that query is at that moment waiting on the same lock, so cannot be canceled. To allow this cancellation `CancelableSharedMutex` is used instead of the regular `std::shared_mutex`. This allows `cancelQuery()` to signal every thread of the query and force blocked `CancelableSharedMutex::lock()` and `CancelableSharedMutex::shared_lock()` to throw exceptions instead of dead-locking.

A similar change of mutex type can be applied to other places with possible deadlocks after we prove this solution to work.

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix deadlocks caused by overcommit tracker in some corner cases

### CODE REVIEW

**Comment by serxa** [comment]
Stateless tests (debug, s3 storage) [1/6] — 00988_parallel_parts_removal FAIL.
Looks like thread cancelations works. Exception is thrown from expected location, but it is unclear why thread was actually canceled.
```
2023-01-26 01:34:05 [b10ced89e595] 2023.01.26 01:33:59.693626 [ 976 ] {e7866061-ecec-4311-8874-b5d25057f838} <Error> executeQuery: Code: 677. DB::Exception: Thread was canceled: While executing MergeTreeThread. (THREAD_WAS_CANCELED) (version 23.1.1.1) (from [::1]:59620) (comment: 00988_parallel_parts_removal.sql) (in query: SELECT count(), sum(x) FROM mt;), Stack trace (when copying this message, always include the lines below):
2023-01-26 01:34:05 
2023-01-26 01:34:05 0. /build/build_docker/../contrib/llvm-project/libcxx/include/exception:134: std::exception::capture() @ 0x18e42b42 in /usr/bin/clickhouse
2023-01-26 01:34:05 1. /build/build_docker/../contrib/llvm-project/libcxx/include/exception:112: std::exception::exception[abi:v15000]() @ 0x18e42b0d in /usr/bin/clickhouse
2023-01-26 01:34:05 2. /build/build_docker/../contrib/poco/Foundation/src/Exception.cpp:27: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, int) @ 0x308409c0 in /usr/bin/clickhouse
2023-01-26 01:34:05 3. /build/build_docker/../src/Common/Exception.cpp:89: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x2157622e in /usr/bin/clickhouse
2023-01-26 01:34:05 4. /build/build_docker/../src/Common/Exception.h:64: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&&, int, bool) @ 0x18e39e4a in /usr/bin/clickhouse
2023-01-26 01:34:05 5. /build/build_docker/../src/Common/Exception.h:72: DB::Exception::Exception<char const (&) [20], void>(int, char const (&) [20]) @ 0x1954da1e in /usr/bin/clickhouse
2023-01-26 01:34:05 6. /build/build_docker/../src/Common/CancelToken.cpp:147: DB::CancelToken::raise() @ 0x2170b3c8 in /usr/bin/clickhouse
2023-01-26 01:34:05 7. /build/build_docker/../src/Common/CancelableSharedMutex.cpp:0: DB::CancelableSharedMutex::lock_shared() @ 0x28bac47a in /usr/bin/clickhouse
2023-01-26 01:34:05 8. /build/build_docker/../contrib/llvm-project/libcxx/include/shared_mutex:331: std::__1::shared_lock<DB::CancelableSharedMutex>::shared_lock[abi:v15000](DB::CancelableSharedMutex&) @ 0x28ba4c67 in /usr/bin/clickhouse
2023-01-26 01:34:05 9. /build/build_docker/../src/Disks/ObjectStorages/MetadataStorageFromDisk.cpp:104: DB::MetadataStorageFromDisk::readMetadata(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) const @ 0x28ba1c09 in /usr/bin/clickhouse
2023-01-26 01:34:05 10. /build/build_docker/../src/Disks/ObjectStorages/MetadataStorageFromDisk.cpp:56: DB::MetadataStorageFromDisk::getFileSize(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) const @ 0x28ba1b7d in /usr/bin/clickhouse
2023-01-26 01:34:05 11. /build/build_docker/../src/Disks/ObjectStorages/DiskObjectStorage.cpp:205: DB::DiskObjectStorage::getFileSize(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) const @ 0x28b120ca in /usr/bin/clickhouse
2023-01-26 01:34:05 12. /build/build_docker/../src/Storages/MergeTree/DataPartStorageOnDisk.cpp:120: DB::DataPartStorageOnDisk::getFileSize(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) const @ 0x2abf82f9 in /usr/bin/clickhouse
2023-01-26 01:34:05 13. /build/build_docker/../src/Storages/MergeTree/MergeTreeMarksLoader.cpp:99: DB::MergeTreeMarksLoader::loadMarksImpl() @ 0x2af57dc6 in /usr/bin/clickhouse
2023-01-26 01:34:05 14. /build/build_docker/../src/Storages/MergeTree/MergeTreeMarksLoader.cpp:157: DB::MergeTreeMarksLoader::loadMarks()::$_0::operator()() const @ 0x2af59018 in /usr/bin/clickhouse
2023-01-26 01:34:05 15. /build/build_docker/../src/Common/CacheBase.h:123: std::__1::pair<std::__1::shared_ptr<DB::MarksInCompressedFile>, bool> DB::CacheBase<wide::integer<128ul, unsigned int>, DB::MarksInCompressedFile, UInt128TrivialHash, DB::MarksWeightFunction>::getOrSet<DB::MergeTreeMarksLoader::loadMarks()::$_0&>(wide::integer<128ul, unsigned int> const&, DB::MergeTreeMarksLoader::loadMarks()::$_0&) @ 0x2af58d8f in /usr/bin/cl
```

**Comment by serxa** [comment]
I have added query cancelation using QueryStatus::cancelQuery() into `OvercommitTracker` just after it picks the query to cancel. So this is now the second way it can kill a query besides throwing an exception. And this can be done by other queries as well, so we do not have to wait for memory allocation by the picked query.

Note that cancelQuery() now accepts code/msg and passes them to CancelToken::signal() for every thread of the query using thread group, effectively unblocking them if they were waiting on a cancelable synchronization primitive (now we have only CancelableSharedMutex).

But, as tests suggest, not everything is fine. Threads are reused for different jobs, and our thread pools are recursive. Now cancel signal can cross job boundary and cancel wrong job. Looks like better integration of CancelToken and ThreadGroup is required for proper work.

**Comment by serxa** [comment]
I reworked how threads became cancelable. Basically, instead of creating `Cancelable` scope while the thread executes a cancelable job, now we reuse the existing infrastructure of `ThreadGroupStatus`. Namely, when a thread joins any group it became cancelable. On entry, it resets any previously saved cancelation signal to avoid the problem we had before: signal could cross job boundaries due to the recursive nature of thread pools. Now thread should never be included in multiple groups.

Also note that if the canceled query will produce more threads, these threads will be canceled as well on entry into the canceled group. But they would not throw exceptions on entry. The entry point is not considered to be a cancelation point. This is done to be able to better control where exceptions could possibly be thrown.

File: utils/check-style/aspell-ignore/en/aspell-dict.txt [hunk 1]
```diff
@@ -531,3 +531,6 @@ zkcopy
 zlib
 znodes
 zstd
+cancelation
```
Comment (serxa) [inline utils/check-style/aspell-ignore/en/aspell-dict.txt:534]: What I have done wrong? Style check still complains: cancelation -> cancellation

**Comment by serxa** [comment]
02294_overcommit_overflow still crashes the server, investigating...

<details>
<summary>BackTrace</summary>
<pre>
########################################
(version 23.2.1.1, build id: 3A2FB41766C1B8B74BBA56FF9786D175219E3613) (from thread 9500) (query_id: 9618051c-d040-4352-ab4b-a021fe4a156d) (query: SELECT number FROM numbers(130000) GROUP BY number SETTINGS max_memory_usage_for_user=5000000,memory_overcommit_ratio_denominator=2000000000000000000,memory_usage_overcommit_max_wait_microseconds=500) Received signal sanitizer trap (-3)
Sanitizer trap.
Stack trace: 0x2280e863 0x22b07456 0x18190a76 0x181a0c33 0x2eb7fd23 0x22800d3c 0x227fa364 0x227fa1da 0x227fa1da 0x2de55a4a 0x2de676e2 0x30919d7a 0x30915c06 0x3054367f 0x30532d9b 0x305343ff 0x228fdbba 0x229016e4 0x7f6c247f0609 0x7f6c24715133
0.1. inlined from ./build_docker/../src/Common/StackTrace.cpp:334: StackTrace::tryCapture()
0. ./build_docker/../src/Common/StackTrace.cpp:295: StackTrace::StackTrace() @ 0x2280e863 in /usr/bin/clickhouse
1. ./build_docker/../src/Daemon/BaseDaemon.cpp:0: sanitizerDeathCallback() @ 0x22b07456 in /usr/bin/clickhouse
2. __sanitizer::Die() @ 0x18190a76 in /usr/bin/clickhouse
3. ? @ 0x181a0c33 in /usr/bin/clickhouse
4.1. inlined from ./build_docker/../src/Interpreters/ProcessList.cpp:0: operator()
4.2. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/__functional/invoke.h:394: decltype(std::declval<DB::UserOvercommitTracker::pickQueryToExclude(MemoryTracker*)::$_6&>()()) std::__1::__invoke[abi:v15000]<DB::UserOvercommitTracker::pickQueryToExclude(MemoryTracker*)::$_6&>(DB::UserOvercommitTracker::pickQueryToExclude(MemoryTracker*)::$_6&)
4.3. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/__functional/invoke.h:479: void std::__1::__invoke_void_return_wrapper<void, true>::__call<DB::UserOvercommitTracker::pickQueryToExclude(MemoryTracker*)::$_6&>(DB::UserOvercommitTracker::pickQueryToExclude(MemoryTracker*)::$_6&)
4.4. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/__functional/function.h:235: std::__1::__function::__default_alloc_func<DB::UserOvercommitTracker::pickQueryToExclude(MemoryTracker*)::$_6, void ()>::operator()[abi:v15000]()
4. ./build_docker/../contrib/llvm-project/libcxx/include/__functional/function.h:716: void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::UserOvercommitTracker::pickQueryToExclude(MemoryTracker*)::$_6, void ()>>(std::__1::__function::__policy_storage const*) @ 0x2eb7fd23 in /usr/bin/clickhouse
5. ./build_docker/../src/Common/OvercommitTracker.cpp:72: OvercommitTracker::needToStopQuery(MemoryTracker*, MemoryTracker*, long) @ 0x22800d3c in /usr/bin/clickhouse
6. ./build_docker/../src/Common/MemoryTracker.cpp:254: MemoryTracker::allocImpl(long, bool, MemoryTracker*) @ 0x227fa364 in /usr/bin/clickhouse
7. ./build_docker/../src/Common/MemoryTracker.cpp:317: MemoryTracker::allocImpl(long, bool, MemoryTracker*) @ 0x227fa1da in /usr/bin/clickhouse
8. ./build_docker/../src/Common/MemoryTracker.cpp:317: MemoryTracker::allocImpl(long, bool, MemoryTracker*) @ 0x227fa1da in /usr/bin/clickhouse
9. ./build_docker/../src/Interpreters/Aggregator.cpp:0: DB::Aggregator::checkLimits(unsigned long, bool&) const @ 0x2de55a4a in /usr/bin/clickhouse
10. ./build_docker/../src/Interpreters/Aggregator.cpp:0: DB::Aggregator::executeOnBlock(std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn>>>, unsigned long, unsigned long, DB::AggregatedDataVariants&, std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*>>&, std::__1::vector<std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*>>, std::__1::allocator<std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*>>>>&, bool&) const @ 0x2de676e2 in /usr/bin/clickhouse
11. ./build_docker/../src/Processors/Transforms/AggregatingTransform.cpp:0: DB::AggregatingTransform::consume(DB::Chunk) @ 0x30919d7a in /usr/bin/clickhouse
12.1. inlined from ./build_docker/../src/Processors/Chunk.h:32: ~Chunk
12. ./build_docker/../src/Processors/Transforms/AggregatingTransform.cpp:492: DB::AggregatingTransform::work() @ 0x30915c06 in /usr/bin/clickhouse
13.1. inlined from ./build_docker/../src/Processors/Executors/ExecutionThreadContext.cpp:50: DB::executeJob(DB::ExecutingGraph::Node*, DB::ReadProgressCallback*)
13. ./build_docker/../src/Processors/Executors/ExecutionThreadContext.cpp:92: DB::ExecutionThreadContext::executeTask() @ 0x3054367f in /usr/bin/clickhouse
14. ./build_docker/../src/Processors/Executors/PipelineExecutor.cpp:229: DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) @ 0x30532d9b in /usr/bin/clickhouse
15.1. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:833: operator()
15.2. inlined from ./build_docker/../base/base/scope_guard.h:99: BasicScopeGuard<DB::PipelineExecutor::spawnThreads()::$_0::operator()() const::'lambda'()>::invoke()
15.3. inlined from ./build_docker/../base/base/scope_guard.h:48: ~BasicScopeGuard
15.4. inlined from ./build_docker/../src/Processors/Executors/PipelineExecutor.cpp:328: operator()
15.5. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/__functional/invoke.h:394: decltype(std::declval<DB::PipelineExecutor::spawnThreads()::$_0&>()()) std::__1::__invoke[abi:v15000]<DB::PipelineExecutor::spawnThreads()::$_0&>(DB::PipelineExecutor::spawnThreads()::$_0&)
15.6. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/tuple:1789: decltype(auto) std::__1::__apply_tuple_impl[abi:v15000]<DB::PipelineExecutor::spawnThreads()::$_0&, std::__1::tuple<>&>(DB::PipelineExecutor::spawnThreads()::$_0&, std::__1::tuple<>&, std::__1::__tuple_indices<>)
15.7. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/tuple:1798: decltype(auto) std::__1::apply[abi:v15000]<DB::PipelineExecutor::spawnThreads()::$_0&, std::__1::tuple<>&>(DB::PipelineExecutor::spawnThreads()::$_0&, std::__1::tuple<>&)
15.8. inlined from ./build_docker/../src/Common/ThreadPool.h:210: operator()
15.9. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/__functional/invoke.h:394: decltype(std::declval<DB::PipelineExecutor::spawnThreads()::$_0>()()) std::__1::__invoke[abi:v15000]<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::PipelineExecutor::spawnThreads()::$_0>(DB::PipelineExecutor::spawnThreads()::$_0&&)::'lambda'()&>(DB::PipelineExecutor::spawnThreads()::$_0&&)
15.10. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/__functional/invoke.h:479: void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::PipelineExecutor::spawnThreads()::$_0>(DB::PipelineExecutor::spawnThreads()::$_0&&)::'lambda'()&>(ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::PipelineExecutor::spawnThreads()::$_0>(DB::PipelineExecutor::spawnThreads()::$_0&&)::'lambda'()&)
15.11. inlined from ./build_docker/../contrib/llvm-project/libcxx/include/__functional/function.h:235: std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::PipelineExecutor::spawnThreads()::$_0>(DB::PipelineExecutor::spawnThreads()::$_0&&)::'lambda'(), void ()>::operator()[abi:v15000]()
15. ./build_docker/../contrib/llvm-project/libcxx/include/__functional/function.h:716: void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::PipelineExecutor::spawnThreads()::$_0>(DB::PipelineExecutor::spawnThreads()::$_0&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) @ 0x305343ff in /usr/bin/clickhouse
16.1. inlined from ./build_docker/../base/base/strong_typedef.h:23: StrongTypedef<std::__1::integral_constant<bool, true> >
16.2. inlined from ./build_docker/../src/Common/OpenTelemetryTraceContext.h:39: DB::OpenTelemetry::Span::isTraceEnabled() const
16. ./build_docker/../src/Common/ThreadPool.cpp:318: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x228fdbba in /usr/bin/clickhouse
17. ./build_docker/../src/Common/ThreadPool.cpp:0: void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x229016e4 in /usr/bin/clickhouse
18. ? @ 0x7f6c247f0609 in ?
19. clone @ 0x7f6c24715133 in ?
</pre>
</details>

**Comment by serxa** [comment]
I missed the case when `pickQueryToExclude()` was unable to select any query. Fixed it. This should fix sanitizer errors we have here now.

**Comment by serxa** [comment]
Now the problem is [here](https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/Transforms/buildPushingToViewsChain.cpp#L277-L292).
For each MV we create special `ThreadStatus`es that are used mainly for metrics of different MVs. They are al connected to the same thread. But for some reason can be deleted and thus call `detachQuery()` from wrong thread. This is detected now because I added an assert [here](https://github.com/ClickHouse/ClickHouse/pull/45236/files#diff-f423985f33dd5d63d4755a441bdab81710a9104242ba01306c77cc7751ab7523R360).

**Comment by serxa** [comment]
One more problem discovered. I've added allocation during pickQueryToExclude() and it leads to crash, need to rework this logic

**Comment by serxa** [comment]
And there is also the deadlock from #47460. See the last column in this reverse flamegraph (made from hung_check.log using stackcollapse-gdb.pl):
hangup-overcommit-reverse

**Comment by serxa** [comment]
I fixed the problem with allocation during `pickQueryToExclude()` just as it is done in most of MemoryTracker code, just by using `MemoryTrackerBlockerInThread`. It should work, but I do not like this, it is better to avoid any allocations at all. But looks like we have a lot of such places already.

Also, there is a new super-unlikely deadlock. `CancelTokenGroup` and `CancelToken` allocate memory under mutex, which could now lead to deadlock in case we got OOM on that allocation. Because during `queryCancel()` (which is now called by OvercommitTracker on OOM) we also need this mutex to signal every thread in group. The most annoying is global mutex `CancelToken::signal_mutex` under which we do std::string allocation `exception_message = message;`. So OOM during cancel or KILL query will lead to deadlock.

I consider this situation very unlikely, so it can be left as is for a while. Eventually, we should rewrite cancel tokens without allocations. Maybe just use `std::pmr::string` over the globally preallocated buffer to store the exception strings of queries being canceled. few KBs should be enough.

**Comment by serxa** [comment]
After fixing many other problems, assert during MV processing shows up again: `Logical error: 'CancelToken::local().thread_id == thread_id'.`
```
4. raise @ 0x7f0c6ca9700b in ?
5. abort @ 0x7f0c6ca76859 in ?
6. /build/src/Common/Exception.cpp:41: DB::abortOnFailedAssertion(String const&) @ 0x22c19936 in /usr/bin/clickhouse
7. /build/src/Interpreters/ThreadStatusExt.cpp:250: DB::ThreadStatus::detachFromGroup() @ 0x2bad03c7 in /usr/bin/clickhouse
8. /build/src/Interpreters/ThreadStatusExt.cpp:292: DB::ThreadStatus::attachToGroup(std::shared_ptr<DB::ThreadGroup> const&, bool)::$_0::operator()() const @ 0x2bad3758 in /usr/bin/clickhouse
9. /build/contrib/llvm-project/libcxx/include/__functional/invoke.h:394: decltype(std::declval<DB::ThreadStatus::attachToGroup(std::shared_ptr<DB::ThreadGroup> const&, bool)::$_0&>()()) std::__invoke[abi:v15000]<DB::ThreadStatus::attachToGroup(std::shared_ptr<DB::ThreadGroup> const&, bool)::$_0&>(DB::ThreadStatus::attachToGroup(std::shared_ptr<DB::ThreadGroup> const&, bool)::$_0&) @ 0x2bad3735 in /usr/bin/clickhouse
10. /build/contrib/llvm-project/libcxx/include/__functional/invoke.h:480: void std::__invoke_void_return_wrapper<void, true>::__call<DB::ThreadStatus::attachToGroup(std::shared_ptr<DB::ThreadGroup> const&, bool)::$_0&>(DB::ThreadStatus::attachToGroup(std::shared_ptr<DB::ThreadGroup> const&, bool)::$_0&) @ 0x2bad3715 in /usr/bin/clickhouse
```
Here is the full [backtrace](https://pastila.nl/?008b48cf/12df315960c90d16fe52ebee6f56359a) from the report.

**Comment by robot-clickhouse-ci-2** [comment]
This is an automated comment for commit 452d095e2b87783d05de09c41a2551ad26015d68 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/45236/452d095e2b87783d05de09c41a2551ad26015d68/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by alexey-milovidov** [comment]
@serxa what is remaining in this wonderful pull request?

**Comment by serxa** [comment]
It is currently not working properly with MV handlers that are called from the same thread with different ThreadStatus.

**Pull request closed without merging**